src/client/dfs/dfs_sys.c

```diff
@@ -0,0 +1,1453 @@
+	 * <dir> + '\0' + <name> + '\0'
+	 */
+	D_ALLOC(new_path, end_idx + 3);
+	if (new_path == NULL) {
+		return ENOMEM;
+	}
+
+	/** Copy the dirname */
+	if (slash_idx == 0)
+		dir_name_len = 1;
+	else
+		dir_name_len = slash_idx;
+	strncpy(new_path, path, dir_name_len);
+	dir_name = new_path;
+
+	/** Copy the basename */
+	name_len = end_idx - slash_idx;
+	if (name_len > 0) {
+		new_path[dir_name_len] = 0;
+		strncpy(new_path + dir_name_len + 1, path + slash_idx + 1,
+			name_len);
+		name = new_path + dir_name_len + 1;
+	}
+
+	sys_path->path = path;
+	sys_path->path_len = path_len;
+	sys_path->alloc = new_path;
+	sys_path->dir_name = dir_name;
+	sys_path->dir_name_len = dir_name_len;
+	sys_path->name = name;
+	sys_path->name_len = name_len;
+
+	rc = hash_lookup(dfs_sys, sys_path);
+	if (rc != 0) {
+		sys_path_free(dfs_sys, sys_path);
+		return rc;
+	}
+
+	return rc;
+}
+
+int
+dfs_sys_mount(daos_handle_t poh, daos_handle_t coh, int mflags, int sflags,
+	      dfs_sys_t **_dfs_sys)
+{
+	dfs_sys_t	*dfs_sys;
+	int		rc;
+	uint32_t	hash_feats = D_HASH_FT_EPHEMERAL;
+	bool		no_cache = false;
+	bool		no_lock = false;
+
+	if (_dfs_sys == NULL)
+		return EINVAL;
+
+	if (sflags & DFS_SYS_NO_CACHE) {
+		D_DEBUG(DB_TRACE, "dfs_sys_mount(): DFS_SYS_NO_CACHE.\n");
+		no_cache = true;
+		sflags &= ~DFS_SYS_NO_CACHE;
+	}
+	if (sflags & DFS_SYS_NO_LOCK) {
+		D_DEBUG(DB_TRACE, "dfs_sys_mount(): DFS_SYS_NO_LOCK.\n");
+		no_lock = true;
+		sflags &= ~DFS_SYS_NO_LOCK;
+	}
+
+	if (sflags != 0) {
+		D_DEBUG(DB_TRACE, "DFS_SYS mount invalid sflags.\n");
+		return EINVAL;
+	}
+
+	D_ALLOC_PTR(dfs_sys);
+	if (dfs_sys == NULL)
+		return ENOMEM;
+
+	/* Mount dfs */
+	rc = dfs_mount(poh, coh, mflags, &dfs_sys->dfs);
+	if (rc != 0) {
+		D_DEBUG(DB_TRACE, "dfs_mount() failed (%d)\n", rc);
+		D_GOTO(err_dfs_sys, rc);
+	}
+
+	/* Initialize the hash */
+	if (!no_cache) {
+		if (no_lock)
+			hash_feats |= D_HASH_FT_NOLOCK;
+		else
+			hash_feats |= D_HASH_FT_RWLOCK;
+
+		rc = d_hash_table_create(hash_feats, DFS_SYS_HASH_SIZE,
+					 NULL, &hash_hdl_ops,
+					 &dfs_sys->hash);
+		if (rc != 0) {
+			D_DEBUG(DB_TRACE, "d_hash_table_create() failed: "
+				DF_RC"\n", DP_RC(rc));
+			D_GOTO(err_hash, rc = daos_der2errno(rc));
+		}
+	}
+
+	*_dfs_sys = dfs_sys;
+	return rc;
+
+err_hash:
+	dfs_umount(dfs_sys->dfs);
+err_dfs_sys:
+	D_FREE(dfs_sys);
+	return rc;
+}
+
+/**
+ * Unmount a file system with dfs_mount and destroy the hash.
+ */
+int
+dfs_sys_umount(dfs_sys_t *dfs_sys)
+{
+	int		rc;
+	d_list_t	*rlink;
+
+	if (dfs_sys == NULL)
+		return EINVAL;
+
+	if (dfs_sys->hash != NULL) {
+		/* Decrease each reference by one. */
+		while (1) {
+			rlink = d_hash_rec_first(dfs_sys->hash);
+			if (rlink == NULL)
+				break;
+
+			d_hash_rec_decref(dfs_sys->hash, rlink);
+		}
+
+		rc = d_hash_table_destroy(dfs_sys->hash, false);
+		if (rc != 0) {
+			D_DEBUG(DB_TRACE, "d_hash_table_destroy() failed: "
+				DF_RC"\n", DP_RC(rc));
+			D_GOTO(err_hash, rc = daos_der2errno(rc));
```

Comment:
Should we still call dfs_umount if hash destroy fails?

Reply:
good question.. i think so, yes.
at least we can match like daos_fini() ,dfs_umount() etc.

Reply:
For some calls, it looks like daos_fini exits, and for others it just logs a message. It's possible that resources may leak on some error conditions:
https://github.com/daos-stack/daos/blob/191145dba81d871f931ab0a733078a2784f203f4/src/client/api/init.c#L248-L262

For the most completeness, I modified dfs_sys_umount to always make each cleanup call, but return the rc of the first FAILED call. This way, for example, if dfs_umount fails, the rc doesn't clobber a failed rc from d_hash_table_destroy.

What do you think?